File: language_construct_performance/bun/scripts/value_obj_minimal_variable_properties.ts
import { NUM_ENTRIES, benchmarkStats, saveStats } from "./benchmark_utils";

// --- VALUE OBJECT CLASSES ---

class MinimalValueObjectNode {
  label: string;
  id: string;
  accessTo: string[];
  details: any;

  edgeTo?: string[];

  constructor(label: string, id: string, accessTo: string[], details: any, edgeTo?: string[]) {
    this.label = label;
    this.id = id;
    this.accessTo = accessTo;
    this.details = details;

    if (edgeTo !== undefined) this.edgeTo = edgeTo;
  }
}

// --- FACTORIES ---

function createValueObject(index: number) {
  const edgeTo = index % 2 === 0 ? ["r.392", "r.40", "r.41", "update", "administrator", "create", "delete", "read"] : undefined;
  const ut = index % 3 === 0 ? 2 : undefined;
  const ag = index % 4 === 0 ? {
    s: { t: 167 },
    a: { t: 3187978, s: { t: 3187978, s: 3149311, r: 42506 } }
  } : undefined;

  const details: any = {
    provider: "aws",
    accountId: "568709751681",
    principal: true,
    tags: ["aKIAYI2NaRQPOT", "dev testing local"],
    mfas: "",
    la: 1772454942 + (index % 1000),
    s: 1,
    cpd: 0,
    pcb: "-",
    lld: 0,
    cd: 1763097939000,
    cb: "-",
    ub: "-",
    ud: 0,
    ua: 1772526871591
  };

  if (ut !== undefined) details.ut = ut;
  if (ag !== undefined) details.ag = ag;

  return new MinimalValueObjectNode(
    "user-dev-test-" + index,
    "u." + (16406 + index),
    ["s.[s3].UACDR", "a.[s3].DARC", "s.[secretsmanager].RACDU", "s.[dynamodb].RCDAU"],
    details,
    edgeTo
  );
}

// --- BENCHMARK RUNNER ---

const stats: Record<string, any> = {};

async function runBenchmark() {
  console.log(`Starting Value Object Variable Properties Minimal Benchmark with ${NUM_ENTRIES.toLocaleString()} entries...`);

  let valueObjArray: MinimalValueObjectNode[] | null = new Array(NUM_ENTRIES);

  console.log("\n--- CREATION ---");
  benchmarkStats("Value Object Minimal Creation", stats, "creationTimeMs", () => {
    for (let i = 0; i < NUM_ENTRIES; i++) {
      valueObjArray![i] = createValueObject(i);
    }
  }, true);

  console.log("\n--- PLAIN TRAVERSAL ---");
  benchmarkStats("Plain Traversal (Value Object Minimal)", stats, "plainTraversalTimeMs", () => {
    let dummyCount = 0;
    for (let i = 0; i < NUM_ENTRIES; i++) {
      if (valueObjArray![i]) continue;
    }
    return dummyCount;
  });

  console.log("\n--- PROPERTY ACCESS ---");
  benchmarkStats("Property Access (Value Object Minimal)", stats, "propAccessTimeMs", () => {
    let sum = 0;
    for (let i = 0; i < NUM_ENTRIES; i++) {
      sum += valueObjArray![i].details?.ag?.a?.s?.r || 0;
    }
    return sum;
  });

  console.log("\n--- FILTERING ---");
  benchmarkStats("Filtering (Value Object Minimal)", stats, "filterTimeMs", () => {
    let matched = 0;
    for (let i = 0; i < NUM_ENTRIES; i++) {
      if (valueObjArray![i].details.la > 1772455500) matched++;
    }
    return matched;
  });

  console.log("\n--- MUTATION ---");
  benchmarkStats("Mutation (Value Object Minimal)", stats, "mutationTimeMs", () => {
    for (let i = 0; i < NUM_ENTRIES; i++) {
      valueObjArray![i].details.la += 1;
    }
  });

  console.log("\n--- DELETE PROPERTY ---");
  benchmarkStats("Delete Property (Value Object Minimal)", stats, "deletePropertyTimeMs", () => {
    for (let i = 0; i < NUM_ENTRIES; i++) {
      delete valueObjArray![i].details.ud;
    }
  });

  // Clear memory
  valueObjArray.length = 0;
  valueObjArray = null;
  Bun.gc(true);

  // Save Stats
  saveStats("stats_variable.json", "value object minimal", stats);
}

runBenchmark().catch(console.error);

File: language_construct_performance/rust/src/bin/value_obj_minimal_variable_properties.rs
use benchmark::benchmark_utils::{benchmark_stats, get_num_entries, save_stats, TrackingAllocator};
use serde_json::{json, Map, Value};
use std::collections::HashMap;

#[global_allocator]
static GLOBAL: TrackingAllocator = TrackingAllocator::new();

#[derive(Clone)]
pub struct ValueObjectNodeMinimalVar {
    pub label: String,
    pub id: String,
    pub access_to: Vec<String>,
    pub details: HashMap<String, Value>,
    pub edge_to: Option<Vec<String>>,
}

fn create_value_object_minimal_variable(index: usize) -> ValueObjectNodeMinimalVar {
    let edge_to = if index % 2 == 0 {
        Some(vec!["r.392".to_string(), "r.40".to_string(), "r.41".to_string(), "update".to_string(), "administrator".to_string(), "create".to_string(), "delete".to_string(), "read".to_string()])
    } else {
        None
    };

    let mut details = HashMap::new();
    details.insert("provider".to_string(), json!("aws"));
    details.insert("accountId".to_string(), json!("568709751681"));
    details.insert("principal".to_string(), json!(true));
    details.insert("tags".to_string(), json!(vec!["aKIAYI2NaRQPOT", "dev testing local"]));
    details.insert("mfas".to_string(), json!(""));
    details.insert("la".to_string(), json!(1772454942 + (index as i64 % 1000)));
    details.insert("s".to_string(), json!(1));
    details.insert("cpd".to_string(), json!(0));
    details.insert("pcb".to_string(), json!("-"));
    details.insert("lld".to_string(), json!(0));
    details.insert("cd".to_string(), json!(1763097939000_i64));
    details.insert("cb".to_string(), json!("-"));
    details.insert("ub".to_string(), json!("-"));
    details.insert("ud".to_string(), json!(0));
    details.insert("ua".to_string(), json!(1772526871591_i64));

    if index % 3 == 0 {
        details.insert("ut".to_string(), json!(2));
    }

    if index % 4 == 0 {
        let mut s_ag = Map::new();
        s_ag.insert("t".to_string(), json!(167));

        let mut s_a = Map::new();
        s_a.insert("t".to_string(), json!(3187978));
        s_a.insert("s".to_string(), json!(3149311));
        s_a.insert("r".to_string(), json!(42506));

        let mut a = Map::new();
        a.insert("t".to_string(), json!(3187978));
        a.insert("s".to_string(), json!(s_a));

        let mut ag = Map::new();
        ag.insert("s".to_string(), json!(s_ag));
        ag.insert("a".to_string(), json!(a));

        details.insert("ag".to_string(), json!(ag));
    }

    ValueObjectNodeMinimalVar {
        label: format!("user-dev-test-{}", index),
        id: format!("u.{}", 16406 + index),
        access_to: vec!["s.[s3].UACDR".to_string(), "a.[s3].DARC".to_string(), "s.[secretsmanager].RACDU".to_string(), "s.[dynamodb].RCDAU".to_string()],
        details,
        edge_to,
    }
}

fn main() {
    let num_entries = get_num_entries();
    let mut stats = Map::new();
    println!("Starting Value Object Minimal Variable Properties Benchmark with {} entries...", num_entries);

    let mut value_obj_array: Vec<Option<Box<ValueObjectNodeMinimalVar>>> = Vec::with_capacity(num_entries);
    for _ in 0..num_entries {
        value_obj_array.push(None);
    }

    println!("\n--- CREATION ---");
    benchmark_stats("Value Object Minimal Creation", &mut stats, "creationTimeMs", true, &GLOBAL, || {
        for i in 0..num_entries {
            value_obj_array[i] = Some(Box::new(create_value_object_minimal_variable(i)));
        }
    });

    println!("\n--- PLAIN TRAVERSAL ---");
    benchmark_stats("Plain Traversal (Value Object Minimal)", &mut stats, "plainTraversalTimeMs", false, &GLOBAL, || {
        let mut dummy_count = 0;
        for i in 0..num_entries {
            if value_obj_array[i].is_some() {
                dummy_count += 1;
            }
        }
        dummy_count
    });

    println!("\n--- PROPERTY ACCESS ---");
    benchmark_stats("Property Access (Value Object Minimal)", &mut stats, "propAccessTimeMs", false, &GLOBAL, || {
        let mut sum: i64 = 0;
        for i in 0..num_entries {
            let details = &value_obj_array[i].as_ref().unwrap().details;
            if let Some(ag_val) = details.get("ag") {
                if let Some(ag) = ag_val.as_object() {
                    if let Some(a_val) = ag.get("a") {
                        if let Some(a) = a_val.as_object() {
                            if let Some(s_val) = a.get("s") {
                                if let Some(s) = s_val.as_object() {
                                    if let Some(r_val) = s.get("r") {
                                        if let Some(r) = r_val.as_i64() {
                                            sum += r;
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        sum
    });

    println!("\n--- FILTERING ---");
    benchmark_stats("Filtering (Value Object Minimal)", &mut stats, "filterTimeMs", false, &GLOBAL, || {
        let mut matched = 0;
        for i in 0..num_entries {
            if value_obj_array[i].as_ref().unwrap().details.get("la").unwrap().as_i64().unwrap() > 1772455500 {
                matched += 1;
            }
        }
        matched
    });

    println!("\n--- MUTATION ---");
    benchmark_stats("Mutation (Value Object Minimal)", &mut stats, "mutationTimeMs", false, &GLOBAL, || {
        for i in 0..num_entries {
            let details = &mut value_obj_array[i].as_mut().unwrap().details;
            let la = details.get("la").unwrap().as_i64().unwrap();
            details.insert("la".to_string(), json!(la + 1));
        }
    });

    println!("\n--- DELETE PROPERTY ---");
    benchmark_stats("Delete Property (Value Object Minimal)", &mut stats, "deletePropertyTimeMs", false, &GLOBAL, || {
        for i in 0..num_entries {
            value_obj_array[i].as_mut().unwrap().details.remove("ud");
        }
    });

    drop(value_obj_array);

    save_stats("stats_variable.json", "value object minimal", stats);
}
